src/game/stats.c

```diff
@@ -0,0 +1,329 @@
+                                    m_LevelPickups += PIERRE_ITEMS;
+                                    m_LevelKillables += 1;
+                                }
+
+                                // Add killable if object triggered
+                                if (Stats_IsObjectKillable(item->object_number)
+                                    && !m_KillableItems[idx]) {
+                                    m_KillableItems[idx] = true;
+                                    m_LevelKillables += 1;
+
+                                    // Add mercenary pickups
+                                    if (item->object_number == O_SKATEKID) {
+                                        m_LevelPickups += SKATEKID_ITEMS;
+                                    }
+                                    if (item->object_number == O_COWBOY) {
+                                        m_LevelPickups += COWBOY_ITEMS;
+                                    }
+                                    if (item->object_number == O_BALDY) {
+                                        m_LevelPickups += BALDY_ITEMS;
+                                    }
+                                }
+                            }
+                        } while (!(trigger & END_BIT));
+                        break;
+                    }
+                } while (!(type & END_BIT));
+            }
+        }
+    }
+}
+
+static bool Stats_IsObjectKillable(int32_t obj_num)
+{
+    for (int i = 0; m_KillableObjs[i] != NO_ITEM; i++) {
+        if (m_KillableObjs[i] == obj_num) {
+            return true;
+        }
+    }
+    return false;
+}
+
+void Stats_ObserveRoomsLoad()
+{
+    m_CachedFloorArray =
+        GameBuf_Alloc(g_RoomCount * sizeof(FLOOR_INFO *), GBUF_ROOM_FLOOR);
+    for (int i = 0; i < g_RoomCount; i++) {
+        const ROOM_INFO *current_room_info = &g_RoomInfo[i];
+        int count = current_room_info->y_size * current_room_info->x_size;
+        m_CachedFloorArray[i] =
+            GameBuf_Alloc(count * sizeof(FLOOR_INFO), GBUF_ROOM_FLOOR);
+        memcpy(
+            m_CachedFloorArray[i], current_room_info->floor,
+            count * sizeof(FLOOR_INFO));
+    }
+}
+
+void Stats_ObserveItemsLoad()
+{
+    m_CachedItemCount = g_LevelItemCount;
+}
+
+void Stats_CalculateStats()
+{
+    m_LevelPickups = 0;
+    m_LevelKillables = 0;
+    m_LevelSecrets = 0;
+    memset(&m_KillableItems, 0, sizeof(m_KillableItems));
+
+    if (m_CachedItemCount) {
+        if (m_CachedItemCount > MAX_ITEMS) {
+            LOG_ERROR("Too Many g_Items being Loaded!!");
+            return;
+        }
+
+        for (int i = 0; i < m_CachedItemCount; i++) {
+            ITEM_INFO *item = &g_Items[i];
+
+            if (item->object_number < 0 || item->object_number >= O_NUMBER_OF) {
+                LOG_ERROR(
+                    "Bad Object number (%d) on Item %d", item->object_number,
+                    i);
+            }
```

Comment:
Here we can exit early with `continue;` to avoid a pointless run of the loop below.